nemo_skills/inference/generate_solutions.py

```diff
@@ -133,29 +134,50 @@ def generate_solutions(cfg: GenerateSolutionsConfig):
 
             data_points.append(data_point)
 
-            if len(data_points) == cfg.batch_size:
-                # batch-computing the outputs
-                outputs = llm.generate(
-                    prompts=[prompt.build_string(data_point) for data_point in data_points],
-                    stop_phrases=list(cfg.prompt.stop_phrases),
-                    **asdict(cfg.inference),
-                )
-
-                for output, original_data_point in zip(outputs, data_points):
-                    # to make it easier to follow up with evaluation and limit accidental errors, we are adding
-                    # all of the ground-truth data to the output file alongside the generated solutions
-                    output.update(original_data_point)
-                    if 'error_message' not in output:
-                        output['error_message'] = extract_error_message(output['generation'])
-
-                    if cfg.generation_key != "generation":
-                        output[cfg.generation_key] = output.pop("generation")
-
-                    fout.write(json.dumps(output) + "\n")
-                data_points = []
+            if len(data_points) == cfg.batch_size or idx == len(data) - 1:
+                if cfg.inference.use_batch_api:
+                    # Using batch API for generation
+                    # Accessing the model directly since CodeExecutionWrapper does not support batch_generate
+                    request_metadata = llm.model.batch_generate(
```

Comment:
This is a bit brittle since not all models support batch generate, only openai and we want to be explicit that we are not using code gen wrapper as users might expect it to be used. How about instead we add a new parameter `use_code` that default to False and then if it's False we don't wrap the model in CodeExecutionWrapper. Please also add a `batch_generate` method to the base model that will by default raise NotImplementedError, so that it's clear that this functionality is not implemented when someone tries to use it with models other than openai